libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -481,48 +534,82 @@ void GCS_MAVLINK::handle_mission_request(AP_Mission &mission, mavlink_message_t
         // set auto continue to 1
         ret_packet.autocontinue = 1;     // 1 (true), 0 (false)
 
-        /*
-          avoid the _send() function to save memory, as it avoids
-          the stack usage of the _send() function by using the already
-          declared ret_packet above
-         */
+        ret_packet.command = cmd.id;
+
+        return MAV_MISSION_ACCEPTED;
+}
+
+void GCS_MAVLINK::handle_mission_request(mavlink_message_t *msg)
+{
+        // decode
+        mavlink_mission_request_t packet;
+        mavlink_msg_mission_request_decode(msg, &packet);
+
+        mavlink_mission_item_t ret_packet{};
+
         ret_packet.target_system = msg->sysid;
         ret_packet.target_component = msg->compid;
         ret_packet.seq = packet.seq;
-        ret_packet.command = cmd.id;
+        ret_packet.mission_type = packet.mission_type;
 
+        MAV_MISSION_RESULT result_code;
+
+        switch (packet.mission_type) {
+        case MAV_MISSION_TYPE_MISSION:
+            result_code = handle_mission_request_mission(msg, packet, ret_packet);
+            break;
+        default:
+            result_code = MAV_MISSION_UNSUPPORTED;
+            break;
+        }
+
+        if (result_code != MAV_MISSION_ACCEPTED) {
+            mavlink_msg_mission_ack_send(chan,
+                                         msg->sysid,
+                                         msg->compid,
+                                         result_code,
+                                         packet.mission_type);
+            return;
+        }
+
+        // we already have a filled structure, use it in place of _send:
         _mav_finalize_message_chan_send(chan, 
-                                        MAVLINK_MSG_ID_MISSION_ITEM_INT,
+                                        MAVLINK_MSG_ID_MISSION_ITEM,
                                         (const char *)&ret_packet,
                                         MAVLINK_MSG_ID_MISSION_ITEM_MIN_LEN,
-                                        MAVLINK_MSG_ID_MISSION_ITEM_INT_LEN,
-                                        MAVLINK_MSG_ID_MISSION_ITEM_INT_CRC);
-    } else {
-        // decode
-        mavlink_mission_request_t packet;
-        mavlink_msg_mission_request_decode(msg, &packet);
+                                        MAVLINK_MSG_ID_MISSION_ITEM_LEN,
+                                        MAVLINK_MSG_ID_MISSION_ITEM_CRC);
+}
+
+
+MAV_MISSION_RESULT GCS_MAVLINK::handle_mission_request_mission(const mavlink_message_t *msg, const mavlink_mission_request_t &packet, mavlink_mission_item_t &ret_packet)
+{
+    AP_Mission *mission = AP::mission();
+    if (mission == nullptr) {
+        return MAV_MISSION_UNSUPPORTED;
+    }
 
         if (packet.seq != 0 && // always allow HOME to be read
```

Comment:
Preexsisting so don't worry about it, but this check actually overlaps with the only way that read_cmd_from_storage can fail.